// protractor-html-spec-reporter/built/html-spec-reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var configuration_parser_1 = require("./configuration-parser");
var execution_display_1 = require("./execution-display");
var execution_metrics_1 = require("./execution-metrics");
var HtmlSpecReporter = (function () {
    function HtmlSpecReporter(configuration) {
        this.started = false;
        this.finished = false;
        this.configuration = configuration_parser_1.ConfigurationParser.parse(configuration);
        this.display = new execution_display_1.ExecutionDisplay(this.configuration);
        this.metrics = new execution_metrics_1.ExecutionMetrics();
    }
    HtmlSpecReporter.prototype.jasmineStarted = function (suiteInfo) {
        this.started = true;
        this.metrics.start(suiteInfo);
        this.display.jasmineStarted(suiteInfo);
    };
    HtmlSpecReporter.prototype.jasmineDone = function (runDetails) {
        this.metrics.stop(runDetails);
        this.display.summary(runDetails, this.metrics);
        this.finished = true;
    };
    HtmlSpecReporter.prototype.suiteStarted = function (result) {
        this.display.suiteStarted(result);
    };
    HtmlSpecReporter.prototype.suiteDone = function (result) {
        this.display.suiteDone();
    };
    HtmlSpecReporter.prototype.specStarted = function (result) {
        this.metrics.startSpec();
        this.display.specStarted(result);
    };
    HtmlSpecReporter.prototype.specDone = function (result) {
        this.metrics.stopSpec(result);
        if (result.status === "pending") {
            this.metrics.pendingSpecs++;
            this.display.pending(result);
        }
        else if (result.status === "passed") {
            this.metrics.successfulSpecs++;
            this.display.successful(result);
        }
        else if (result.status === "failed") {
            this.metrics.failedSpecs++;
            this.display.failed(result);
        }
        this.display.specDone(result);
    };
    return HtmlSpecReporter;
}());
exports.HtmlSpecReporter = HtmlSpecReporter;
